### ISSUE TITLE
No test coverage for amp unhandled client-side failures

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2562]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2562|
|Type|defect|
|Created|2007-04-10 13:46:14Z|
|Branch|https://github.com/twisted/twisted/tree/amp-local-errback-test-2562|

This patch causes no tests to fail:

```
Index: twisted/protocols/amp.py
===================================================================
--- twisted/protocols/amp.py    (revision 19972)
+++ twisted/protocols/amp.py    (working copy)
@@ -649,7 +649,6 @@
         """
         if ANSWER in box:
             question = self._outstandingRequests.pop(box[ANSWER])
-            question.addErrback(self._puke)
             question.callback(box)
         elif ERROR in box:
             question = self._outstandingRequests.pop(box[ERROR])
```

I don't really see why that errback is there in the first place, though.

<details><summary>Searchable metadata</summary>

```
trac-id__2562 2562
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_amp_local_errback_test_2562 branches/amp-local-errback-test-2562
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1176212774000000 1176212774000000
changetime__1316781040000000 1316781040000000
version__None None
owner__exarkun exarkun
cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [32488]) Branching to 'amp-local-errback-test-2562'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

I added a few tests.  I changed the style of the error logging, too.  [Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/amp-local-errback-test-2562).

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_4"></a>|@jesstess set owner to @exarkun|
|-|-|

The tests and logging changes seem reasonable and do catch removing the ```addErrback``` line from ```_answerReceived```. It is still the case that commenting out the other ```addErrback``` line, in ```_errorReceived```, causes no tests to fail.

So your changes are strictly an improvement and please merge if you want to be done, or at your option look at the ```_errorReceived``` situation or open a new ticket for it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [32516]) Add a test for the network-error case as well

refs [#2562](https://github.com/twisted/twisted/issues/2562)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

Thanks!  I added a test for the error path as well.  [Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/amp-local-errback-test-2562)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_9"></a>|@jesstess set owner to @exarkun|
|-|-|

Looks great, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set status to `closed`|
|-|-|

(In [32575]) Merge amp-local-errback-test-2562

Author: exarkun
Reviewer: jesstess
Fixes: [#2562](https://github.com/twisted/twisted/issues/2562)

Add unit tests for `twisted.protocols.amp`'s logging of unhandled exceptions
in local callbacks on any Deferred returned from `callRemote`.  Also change the
logging style to emit a single message in this case, rather than two.
